hypervisor/src/arch/x86/emulator/mod.rs

```diff
@@ -389,9 +389,48 @@ impl<'a, T: CpuStateManager> Emulator<'a, T> {
             .cpu_state(cpu_id)
             .map_err(EmulationError::PlatformEmulationError)?;
         let mut decoder = Decoder::new(64, insn_stream, DecoderOptions::NONE);
+        let mut insn = Instruction::default();
+        let mut num_insn_emulated: usize = 0;
+        let mut fetched_insn_stream: [u8; 16] = [0; 16];
+        let mut last_decoded_ip: u64 = state.ip();
+        let mut stop_emulation: bool = false;
+
         decoder.set_ip(state.ip());
 
-        for (index, insn) in &mut decoder.iter().enumerate() {
+        while decoder.can_decode() && !stop_emulation {
+            decoder.decode_out(&mut insn);
+
+            if decoder.last_error() == DecoderError::NoMoreBytes {
+                // The decoder is missing some bytes to decode the current
+                // instruction, for example because the instruction stream
+                // crosses a page boundary.
+                // We fetch 16 more bytes from the instruction segment,
+                // decode and emulate the failing instruction and terminate
+                // the emulation loop.
+                warn!(
```

Comment:
This should perhaps be `debug!`. Nothing wrong with needing to fetch more bytes.